### ISSUE TITLE
xmlHttpRequest - finalUrl returning NULL

### ISSUE THREAD

**Issue by Charles152** [greasemonkey/greasemonkey#2692]
GM.xmlHttpRequest's "finalUrl" is returning NULL on a 302 Found redirect, which is observable in Wireshark:

Wireshark Output:

Hypertext Transfer Protocol:
HTTP/1.1 302 Found\r\n
[...]
Location: (Redirect URL)

Greasemonkey output: 

context : null
finalUrl : null
lengthComputable : null
loaded : null
readyState : 4
response : \<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
(rest of response document)
responseXML : null
status : 200
statusText : OK
total : null

This works without a problem in Tampermonkey. Tampermonkey output: 

readyState : 4
responseHeaders : (list of returned headers)
finalUrl : (non-Null redirect URL)
status : 200
statusText : OK
response : \<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
(rest of response document)
responseXML : [object XMLDocument]
UNSENT : 0
OPENED : 1
HEADERS_RECEIVED : 2
LOADING : 3
DONE : 4

The following identical code is being used in both Greasemonkey and Tampermonkey where the URL variable is replaced for the same valid URL:

The HTTP code generating website "https://httpstat.us/302" should work as a valid test URL for this.

```
var URL = "https://httpstat.us/302";

GM.xmlHttpRequest({
  method: "POST",
  url: URL,
  onload: function(response) {
	try {
	  for(var i in response) {
		console.log(i + " : " + response[i]);
	  }
	}catch(e){
	  console.log(e);
	}
  }
});
```

At first I thought mozAnon was causing the problems (since Cookies being sent will cause a similar problem in the script), but mozAnon is  another issue that seems to be happening at random and the 'finalUrl' being null is what is actually causing the problem with the script.

**Comment by arantius** [comment]
Yep, documentation wasn't updated perfectly; this feature is completely missing in GM4.

**Comment by DoomTay** [comment]
Happens with status code of 200 or 404 as well, like https://httpstat.us/200

**Comment by alexgit2k** [comment]
User-Script-Example showing the bug: [bugdemo.txt](https://github.com/greasemonkey/greasemonkey/files/1491968/bugdemo.txt)

Any workarounds to get the requested url within an event-handler without finalUrl?

**Comment by DoomTay** [comment]
Best workaround I can think of is to reuse the "URL" variable, whether it's the string provided or whatever.

**Comment by arantius** [comment]
Pull #2720 should address this but I'd like to add a unit test before I close the issue.  (Or @alexgit2k you could add it, but I don't think there's any tests of this code at all yet, so it's a tall order.)

**Comment by arantius** [comment]
The GM.xhr implementation, as it is, is tightly coupled to webext APIs in a way that makes it a mess to test.  Pulling in this simple fix for now, tests will be added .. some time.

**Comment by arantius** [comment]
This fix should be in 4.1beta6; testing and confirming this would be very appreciated!

https://addons.mozilla.org/firefox/downloads/file/800957/greasemonkey-4.1beta6-an+fx.xpi?src=devhub

**Comment by alexgit2k** [comment]
Working, thanks.
